### ISSUE TITLE
change from "list of connectors per location" to "list of connectors of EVSEs of location"

### ISSUE THREAD

**Issue by poohsen** [ocpi/ocpi#19]
It is an important piece of information to the user to know how many users can be accommodated in a given location. A flat list of, say, 5 connectors, seems to suggest that 5 cars could be charging at a location in parallel. If those 5 connectors are hooked up to only 2 EVSEs, then only 2 users will be able to charge. It seems that this implicit information hasn't been addressed in the current structure of a charge location. 

We propose to introduce an EVSE level in the charge location structure and move the status and evse_id fields to this level, leaving the rest with the connectors.

Thus, this:

connectors (list)
* evse_id: the unique identifier of the EVSE used to provide the power via this connector
* type: enumeration (see ConnectorType specification in OCPP2.0)
* location: exact location of the connector (optional)
  * address: address of the entry location in order to reach the charge point
  * point: GPS coordinates in order to reach the charge point (e.g. at a parking)
  * floor: String indication of the floor according to the local situation.
* capabilities: enumeration (charging_profile_capable, reservable)
* charge_protocol: enumeration (unknown, mode3, chademo, iso15118, uncontrolled)
* status: enumeration (available, occupied, charging, outofservice)
* power (replaces enum of DC50kWh / AC11kWh etc etc)
  * current (ac_1_phase, ac_2_phase, ac_3_phase, dc)
  * voltage
  * amperage
* price_schemes (list of available options)
  * price_schema_id: unique identifier of this schema (at the operator)
  * display_text: human readable form indicating this scheme (list of)
    * language: ISO0639-1
    * text: String
  * start_date: ISO8601 date from this pricing scheme is valid (inclusive)
  * expiry_date: ISO8601 date until this pricing scheme is valid (inclusive)
  * tariff (list of components that build the price)
    * tariff_id: identifier of this tariff (unique or at least within this schema)
    * validity_rule: period_type (enum: Charging, Parking), time (iCalendar RRULE)
    * display_text: human readable form of this part of the tariff
    * currency: ISO 4217 code for currency
    * pricing_unit: enumeration of types of pricing (kwhtoev, occupancyhours, charginghours, idlehours, session see OCPP2)
    * price_net: amount (in smallest unit for relevant currency with an additional two decimal places, incl. VAT. e.g. euros = 0.2343, Japanese yen = 45.34)
    * price_gross: amount (Price of the unit excluding tax. Calculated as 100 * priceNet / (100 + taxPct).
    * tax_percentage: percentage of tax

becomes

EVSE (list)
* evse_id: the unique identifier of the EVSE used to provide the power via this connector
* status: enumeration (available, occupied, charging, outofservice)

* connectors (list)
  * type: enumeration (see ConnectorType specification in OCPP2.0)
  * location: exact location of the connector (optional)
    * address: address of the entry location in order to reach the charge point
    * point: GPS coordinates in order to reach the charge point (e.g. at a parking)
    * floor: String indication of the floor according to the local situation.
  * capabilities: enumeration (charging_profile_capable, reservable)
  * charge_protocol: enumeration (unknown, mode3, chademo, iso15118, uncontrolled)
  * power (replaces enum of DC50kWh / AC11kWh etc etc)
    * current (ac_1_phase, ac_2_phase, ac_3_phase, dc)
    * voltage
    * amperage
  * price_schemes (list of available options)
    * price_schema_id: unique identifier of this schema (at the operator)
    * display_text: human readable form indicating this scheme (list of)
      * language: ISO0639-1
      * text: String
    * start_date: ISO8601 date from this pricing scheme is valid (inclusive)
    * expiry_date: ISO8601 date until this pricing scheme is valid (inclusive)
    * tariff (list of components that build the price)
      * tariff_id: identifier of this tariff (unique or at least within this schema)
      * validity_rule: period_type (enum: Charging, Parking), time (iCalendar RRULE)
      * display_text: human readable form of this part of the tariff
      * currency: ISO 4217 code for currency
      * pricing_unit: enumeration of types of pricing (kwhtoev, occupancyhours, charginghours, idlehours, session see OCPP2)
      * price_net: amount (in smallest unit for relevant currency with an additional two decimal places, incl. VAT. e.g. euros = 0.2343, Japanese yen = 45.34)
      * price_gross: amount (Price of the unit excluding tax. Calculated as 100 * priceNet / (100 + taxPct).
      * tax_percentage: percentage of tax

**Comment by simonphilips** [comment]
Minor corrections:

(1) EVSE.connectors.location is deleted because this information is now moved to the main Location object and EVSE itself.

The Location object describes the location and its properties where a group of EVSE's that belong together are installed. Typically the Location object is the exact location of the group of EVSE's, but it can also be the entrance of a parking garage which contains these EVSE's. Each EVSE can then be further specified by its floor/directions/physical_number properties.

The idea is that a driver first follows the instructions towards the Location, and any further instructions from the Location the EVSE belong in to the EVSE object through its floor, directions and physical_number fields.

* EVSE.floor: decimal (since half-floors are possible)
* EVSE.directions: text that provides instructions on how to get there from the Location object
* EVSE.physical_number: a number on the EVSE for visual identification within the Location (like at a gas station)

It was decided that having an address/coordinates in both the Location and in the EVSE would be confusing as there is too much information and uncertainty about which one to use/show. When EVSE.floor, EVSE.directions and EVSE.physical_number are insufficient to reach the EVSE from the Location point, then it typically indicates that this EVSE should be put in a different Location object (sometimes with the same address but with different coordinates/directions).


(2) a connector should also have an optional terms_and_conditions url next to the price_schemes.

**Comment by simonphilips** [comment]
I've put the objects and properties as mentioned above into some sort of class diagram. I based the naming on the vocabulary we used during the previous meeting and on what is in the current draft.

![ocpi](https://cloud.githubusercontent.com/assets/3263809/7397564/90bdd350-eea9-11e4-93e9-d6538251983e.png)
(Attributes marked with a green circle are required)

I have also added an extra property "facilities" to specify what kind of facilities are available at a certain location (such as a hotel or a restaurant). It's a feature requested by our customers, we are going to add it to our own protocol but I think it would be interesting to discuss its use for OCPI as well.
